### PR TITLE
Add support for MP3 and FLAC

### LINKED ISSUE #16: Support for other formats
Support for common formats such as MP3, FLAC and OGG.

- [x] WAV
  - [x] dr_wav
  - [ ] custom?
- [x] MP3
  - [x] dr_mp3
- [x] FLAC
  - [x] dr_flac
- [ ] #21

### PR DESCRIPTION
**Pull request by alexdevteam**
- Adds support for MP3 and FLAC
- Showcases it in the example
Closes #16

### CODE REVIEW

File: example/example.cpp [hunk 1]
```diff
@@ -1,17 +1,35 @@
...
 
 namespace impl {
 static constexpr int fail_code = 2;
 // TODO: Add parameters for each one of these constants [Use clap?]
 static constexpr float travel_circurference_radius = 2.0f;
 static constexpr float travel_angular_speed = 2.0f;
-static constexpr bool loop_audio = true;
+static constexpr bool loop_audio = false;
+/* clang-format off */
+static std::vector<std::tuple<std::string, capo::FileFormat>> supportedFormats{
```
Comment (karnkaul) [inline example/example.cpp:20]: This should be constexpr, array, and string_view

File: example/example.cpp [hunk 2]
```diff
@@ -1,17 +1,35 @@
...
+	{".wav", capo::FileFormat::eWav},
+	{".flac", capo::FileFormat::eFlac},
+	{".mp3", capo::FileFormat::eMp3}
+};
+/* clang-format on */
+
+std::optional<capo::FileFormat> formatFromFilename(std::string name) {
```
Comment (karnkaul) [inline example/example.cpp:27]: `name` should be `string_view`

File: include/capo/pcm.hpp [hunk 3]
```diff
@@ -16,6 +17,8 @@ struct SampleMeta {
 	std::uint8_t channels{};
 };
 
+enum class FileFormat { eWav, eMp3, eFlac, eCOUNT_ };
```
Comment (karnkaul) [inline include/capo/pcm.hpp:20]: I'd prefer having the first entry be `eUnknown`, would also eliminate the optional in the other file.

File: include/capo/pcm.hpp [hunk 4]
```diff
@@ -28,7 +31,8 @@ struct PCM {
 	std::vector<Sample> samples;
 	utils::Size size{};
 
-	static Result<PCM> fromFile(std::string const& wavPath);
-	static Result<PCM> fromMemory(std::span<std::byte const> wavBytes);
+	static Result<PCM> fromFile(std::string const& path, FileFormat format);
```
Comment (karnkaul) [inline include/capo/pcm.hpp:34]: Defaulting the second param to unknown would also allow a "deduction" feature (from extension).

File: include/capo/utils/file_utils.hpp [hunk 5]
```diff
@@ -0,0 +1,21 @@
+#pragma once
+#include <fstream>
+#include <string>
+#include <vector>
+
+namespace capo::utils {
+
+inline std::vector<std::byte> fileBytes(std::string const& path) {
```
Comment (karnkaul) [inline include/capo/utils/file_utils.hpp:8]: Where was this before?

File: src/pcm.cpp [hunk 6]
```diff
@@ -1,68 +1,99 @@
+#define DR_FLAC_IMPLEMENTATION
+#include <dr_libs/dr_flac.h>
+#define DR_MP3_IMPLEMENTATION
+#include <dr_libs/dr_mp3.h>
 #define DR_WAV_IMPLEMENTATION
 #include <dr_libs/dr_wav.h>
 
 #include <capo/pcm.hpp>
 #include <algorithm>
+#include <cstring>
```
Comment (karnkaul) [inline src/pcm.cpp:10]: Why is this needed?

File: src/pcm.cpp [hunk 7]
```diff
@@ -1,68 +1,99 @@
...
 #include <fstream>
+#include <iostream>
```
Comment (karnkaul) [inline src/pcm.cpp:12]: IO happening here?

File: src/pcm.cpp [hunk 8]
```diff
@@ -1,68 +1,99 @@
...
 #include <optional>
 
 namespace capo {
 namespace {
-class WAV {
```
Comment (karnkaul) [inline src/pcm.cpp:11]: Why has this been removed?

File: src/pcm.cpp [hunk 9]
```diff
@@ -1,68 +1,99 @@
...
-		ret.meta.rate = static_cast<std::size_t>(wav.m_wav.sampleRate);
-		ret.meta.channels = static_cast<std::uint8_t>(wav.m_wav.channels);
-		ret.size = utils::Size::make(ret.samples);
-		return ret;
+Result<PCM> PCM::fromMemory(std::span<std::byte const> bytes, FileFormat format) {
+	if (bytes.empty()) { return Error::eIOError; }
+
+	static_assert(static_cast<std::size_t>(FileFormat::eCOUNT_) == 3, "Unhandled file format");
+	switch (format) {
+	case FileFormat::eWav: return decode<FileFormat::eWav>(bytes);
```
Comment (karnkaul) [inline src/pcm.cpp:92]: None of these handles are closed anymore?

**Review by karnkaul — CHANGES_REQUESTED** [review]

File: include/capo/utils/file_utils.hpp (hunk 5 above)
Comment (alexdevteam) [reply include/capo/utils/file_utils.hpp:8]: Inside pcm.cpp; I think I had to pull it to make some internal changes but it's no longer required (even then it's not a bad idea to have it in a separate file)

File: src/pcm.cpp (hunk 6 above)
Comment (alexdevteam) [reply src/pcm.cpp:10]: Left from using memcpy, oops

File: src/pcm.cpp (hunk 7 above)
Comment (alexdevteam) [reply src/pcm.cpp:12]: Left from debugging

File: src/pcm.cpp (hunk 8 above)
Comment (alexdevteam) [reply src/pcm.cpp:11]: No real reason to have it previously; there was no need to keep track of the dr_libs objects, since only reading the files fully was required. I decided to simplify it using functions instead.

File: src/pcm.cpp (hunk 9 above)
Comment (alexdevteam) [reply src/pcm.cpp:92]: Oops nope, that should happen inside decode, my bad

File: src/pcm.cpp (hunk 8 above)
Comment (karnkaul) [reply src/pcm.cpp:11]: There's very much reason to have it, lifetime is critically important. It's also needed (features expanded) in streaming, so I'd suggest trying to incorporate a design around it / using it.

Comment (karnkaul) [reply src/pcm.cpp:11]: I'll push the latest changes to that unstable branch so the potential "future" usage is visible.

Comment (alexdevteam) [reply src/pcm.cpp:11]: RAII already handles this. Data is stored and the lifetime is handled by PCM. If an explicit deinit is needed it will occur automatically. The data needed right now is fetched on load so there's no need to hold the dr_libs object at any point in time. And streaming will need a completely different interface anyways because data will be supplied progressively, so there is no reason to keep the old approach in my opinion.

File: include/capo/types.hpp [hunk 10]
```diff
@@ -20,6 +20,7 @@ enum class Error {
 	eDeviceFailure,
 	eContextFailure,
 	eInvalidValue,
+	eUnknownFormat,
```
Comment (karnkaul) [inline include/capo/types.hpp:23]: This will need a corresponding update to the associated array of strings in impl_al.hpp

File: src/pcm.cpp [hunk 11]
```diff
@@ -55,41 +169,33 @@ std::vector<std::byte> fileBytes(std::string const& path) {
...
-		.format = wav.channels == 2 ? SampleFormat::eStereo16 : SampleFormat::eMono16,
-		.channels = static_cast<std::size_t>(wav.channels),
-	};
-}
 } // namespace
 
-bool PCM::supported(SampleMeta const& meta) noexcept {
-	if (meta.channels == 0 || meta.channels > max_channels_v) { return false; }
-	return meta.rate > 0;
+constexpr bool SampleMeta::supported() noexcept {
```
Comment (karnkaul) [inline src/pcm.cpp:174]: I'm surprised this compiles! `constexpr` is usually required to be inline...

File: example/example.cpp [hunk 12]
```diff
@@ -22,32 +24,48 @@ capo::Vec3 UCMVelocity(float angularSpeed, float time, float radius) {
...
+		return false;
 	}
 
 	auto pcm = capo::PCM::fromFile(wavPath);
 	if (!pcm) {
-		std::cerr << "Couldn't load audio file." << std::endl;
-		return impl::fail_code;
+		switch (pcm.error()) {
+		case capo::Error::eUnknownFormat:
+			static_assert(static_cast<std::size_t>(capo::FileFormat::eCOUNT_) == 4, "Unhandled file format");
```
Comment (karnkaul) [inline example/example.cpp:38]: This static assert seems unnecessary...

File: example/example.cpp [hunk 13]
```diff
@@ -22,32 +24,48 @@ capo::Vec3 UCMVelocity(float angularSpeed, float time, float radius) {
...
-		return impl::fail_code;
+		return false;
 	}
 	source.loop(loop);
 	source.play();
 
-	std::cout << ktl::format("Playing {} once at {.2f} gain, length: {.1f}s\n", wavPath, gain, sound.length().count());
-	std::cout << ktl::format("Travelling on a circurference around the listener; r={.1f}, angular speed={.1f}\n", travel_circurference_radius,
-							 travel_angular_speed);
+	std::cout << ktl::format("{} info:\n\t{}s Length\n\t{} Channel(s)\n\t{}Hz Sample Rate\n\t{} Size\n", wavPath, sound.length().count(),
```
Comment (karnkaul) [inline example/example.cpp:60]: Just FYI: `channels` is now plain size_t, so printing it should work normally.

File: include/capo/types.hpp (hunk 10 above)
Comment (alexdevteam) [reply include/capo/types.hpp:23]: Why is there no static assert in EnumArray to warn that?

File: example/example.cpp (hunk 12 above)
Comment (alexdevteam) [reply example/example.cpp:38]: It's tied to the next output message. If we end up supporting more formats we'll need to add them manually to the supported formats cout

File: include/capo/types.hpp (hunk 10 above)
Comment (karnkaul) [reply include/capo/types.hpp:23]: I don't think it's possible to: the array size is based on `eCOUNT_`, it's just that you're initializing less than the total elements.

Comment (karnkaul) [reply include/capo/types.hpp:23]: I'm thinking... We could add a templated constructor, and static_assert that the number of args passed equals the array size. Would that be acceptable?

Comment (alexdevteam) [reply include/capo/types.hpp:23]: Yeah, definitely

Comment (alexdevteam) [reply include/capo/types.hpp:23]: That can be on a separate PR though, I'll add the eUnknownFormat string

**Comment by alexdevteam** [comment]
Please squash the commits if you do end up merging the PR

**Review by karnkaul — APPROVED** [review]
